### ISSUE TITLE
Is there a way to compress the tim files back to LPK?

### ISSUE THREAD

**Issue by shoddydev** [Hedzin/SquareNPK-unpacker#1]
Thank you very much for this. I used it for a different PSX game with .lpk archives and it worked. I was wondering if there's a way to compress the files back to .lpk for modding purposes.

**Comment by Hedzin** [comment]
what game do you mean? i suppose there are tons of different software to edit most popular squaresoft games

**Comment by shoddydev** [comment]
I used your SquareNPK-unpacker for RPG Maker on the PSX. There's a file on the disc named TIM_TT.LPK and it contains the main menu artwork. I'm creating my own RPG with it using modified assets and I would I would need to compress a .tim image back to .lpk to use my own customized menu. I tried using some of the FF7 international LPK files in RPG Maker and it worked.

**Comment by Hedzin** [comment]
I suppose it is possible. 
do you have a kind of small .lpl file to play with it?

**Comment by shoddydev** [comment]
Sure here's the ,lpk from RPG maker. I used the processSmallNPK(NPK_MAIN,3,0x1c800,0x92000); procedure to extract the .tims,
https://drive.google.com/file/d/1IRrDqArs-3-KLUMqoTSw_mpteOPVyfil/view?usp=sharing

Normally I convert a bmp to .tim, then I use a hex editor to insert the replacement tim and then I use cdmage to inject the modified file. I'm pretty sure it'll work I just need a way to repack tim images to a .lpk archive.

**Comment by Hedzin** [comment]
i commited 2 new classes for your lpk.
i did not implement lzss comression as there is tool for it that works:
https://github.com/cebix/ff7tools
so it should work like 
1)unpack file.
2) edit it
3) compress every file with  ff7tools lzss 
4) pack files back with RPGMakerPacker

but there are some notes:
1) you have to delete line 78 from ff7tools lzss before use it
2) result files after ff7tools lzss compression are differend from original, 
despite tha fact that tecnically it should works(if you unpack file comperessed with  ff7tools lzss back with RPGMakerExtractor you will recieve valid .tim files), there is potential possability that original rpgmaker code may failed.

so to test it i advice you to perform full flow with no changes, just:
1) unpack
2)compress
3) pack again
4)test with rpgmaker

to check that rpg maker is ok with reading it

**Comment by shoddydev** [comment]
It almost worked. But the repack appears to be missing the DF 10 00 00 00 09 header, When I first inject it without the header it'll show the image momentarily but then crash. I was able to edit the headers a bit so that it closely matches the original but when displayed in the game the colors are jumbled similar to outline of the sword in your early attempts. It's real close though, if there's a way to repack it to include the DF 10 00 00 00 09 headers I'm sure that'll work.

**Comment by Hedzin** [comment]
its not a header its a difference in lzss implemnentation.
technically files are compressed correctly, but the results are different in original algorithem and by using ff7 tools.

original algorithem  for some reasons leaves first bytes uncompressed

**Comment by Hedzin** [comment]
try to play with lzss compression settings
in file /ff7/lzss.py there are 2 values 
MAX_REF_LEN = 18  # maximum reference length
MIN_REF_LEN = 3   # minimum reference length

try to change it to 5 and 17
9 and 20 
5 and 7

as i remember some variants leaved header little bit more uncompressed

**Comment by shoddydev** [comment]
I finally got it to work! I kept the default settings at 18 and 3 though. Basically in a hex editor I just copied the entire data that FFtools lzss generated, then in the TIM_TT.LPK file I paste it where the DF 10 00 00 00 09 began and overwrite the data. And it worked. Thank you once again, I'm very happy I can replace the Menu background.
![menu_mod](https://user-images.githubusercontent.com/42045893/89436546-40c4dc00-d6fb-11ea-8c31-37e1a6b6c3bf.jpg)
